### PR TITLE
[Merged by Bors] - feat: generalized eigenspaces are independent

### PR DESCRIPTION
**Pull request by ocfnash**
The main change is to upgrade the existing `Module.End.eigenspaces_independent`, which applied only to eigenspaces (and required a `[Field K]` assumption) to `Module.End.independent_generalizedEigenspace`, which applies to generalized eigenspaces (and requires only `[NoZeroSMulDivisors R M]`)

---

Split out from #7963

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
This is the not the definition I was expecting for `IsTriangularizable`. Instead, I was expecting: "there exists a maximal chain of invariant subspaces". Over a finite dimensional vector space over a field, the definitions coincide, but not in the infinite dimensional setting. Thoughts?

**Comment by ocfnash** [comment]
> This is the not the definition I was expecting for `IsTriangularizable`. Instead, I was expecting: "there exists a maximal chain of invariant subspaces". Over a finite dimensional vector space over a field, the definitions coincide, but not in the infinite dimensional setting. Thoughts?

Great point, I should have thought of this myself! I'm already yak shaving to some extent so I'm not sure if I have the appetite to develop the full theory at the moment. A reasonable compromise might be keep these results without taking ownership of the `IsTriangularizable` name.

Given your remarks, I'll split this PR in two: I think the results about independence should be less controversial.

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 1]
```diff
@@ -379,13 +268,136 @@ theorem generalizedEigenspace_le_generalizedEigenspace_finrank [FiniteDimensiona
...
+    (f : End R M) {μ₁ μ₂ : R} (hμ : μ₁ ≠ μ₂) (k l : ℕ) :
+    Disjoint (f.generalizedEigenspace μ₁ k) (f.generalizedEigenspace μ₂ l) := by
+  nontriviality M
+  have := NoZeroSMulDivisors.isReduced R M
+  rw [disjoint_iff]
+  set p := f.generalizedEigenspace μ₁ k ⊓ f.generalizedEigenspace μ₂ l
+  by_contra hp
+  replace hp : Nontrivial p := Submodule.nontrivial_iff_ne_bot.mpr hp
+  have hc₁ : Commute f (f - algebraMap R (End R M) μ₁) :=
+    Commute.sub_right rfl (Algebra.commute_algebraMap_right μ₁ f)
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Eigenspace/Basic.lean:310]: This is such a commonly used fact in linear algebra arguments that I wonder if we should just make it a lemma:
```lean
lemma commute_self_sub_algebraMap (f : End R M) (μ : R) :
    Commute f (f - algebraMap R (End R M) μ) :=
  Commute.sub_right rfl (Algebra.commute_algebraMap_right μ f)
```

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 2]
```diff
@@ -379,13 +268,136 @@ theorem generalizedEigenspace_le_generalizedEigenspace_finrank [FiniteDimensiona
...
+  have hc₂ : Commute f (f - algebraMap R (End R M) μ₂) :=
+    Commute.sub_right rfl (Algebra.commute_algebraMap_right μ₂ f)
+  let f₁ : End R p := LinearMap.restrict (f - algebraMap R (End R M) μ₁) (MapsTo.inter_inter
+    (mapsTo_generalizedEigenspace_of_comm hc₁ μ₁ k) (mapsTo_generalizedEigenspace_of_comm hc₁ μ₂ l))
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Eigenspace/Basic.lean:314]: I think the change the first line makes it more readable, and the remaining lines use the lemma I suggested so you don't need to declare `hc₁`.
```suggestion
  let f₁ : End R p := (f - algebraMap R (End R M) μ₁).restrict <| MapsTo.inter_inter
    (mapsTo_generalizedEigenspace_of_comm (f.commute_self_sub_algebraMap μ₁) μ₁ k)
    (mapsTo_generalizedEigenspace_of_comm (f.commute_self_sub_algebraMap μ₁) μ₂ l)
```
If you still *wanted* to use `hc₁`, I would consider doing it *within* the `let` declaration with `haveI`, as in:
```lean
  let f₁ : End R p := haveI hc := (f.commute_self_sub_algebraMap μ₁)
    (f - algebraMap R (End R M) μ₁).restrict <| MapsTo.inter_inter
      (mapsTo_generalizedEigenspace_of_comm hc μ₁ k) (mapsTo_generalizedEigenspace_of_comm hc μ₂ l)
```
which makes it clear that `hc₁` is only used to define `f₁`.

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean [hunk 3]
```diff
@@ -379,13 +268,136 @@ theorem generalizedEigenspace_le_generalizedEigenspace_finrank [FiniteDimensiona
...
+lemma disjoint_iSup_generalizedEigenspace [NoZeroSMulDivisors R M]
+    (f : End R M) {μ₁ μ₂ : R} (hμ : μ₁ ≠ μ₂) :
+    Disjoint (⨆ k, f.generalizedEigenspace μ₁ k) (⨆ k, f.generalizedEigenspace μ₂ k) := by
+  simp_rw [(f.generalizedEigenspace μ₁).mono.directed_le.disjoint_iSup_left,
+    (f.generalizedEigenspace μ₂).mono.directed_le.disjoint_iSup_right]
+  exact disjoint_generalizedEigenspace f hμ
+
+lemma injOn_generalizedEigenspace [NoZeroSMulDivisors R M] (f : End R M) :
+    InjOn (fun μ ↦ ⨆ k, f.generalizedEigenspace μ k)
+      {μ | ⨆ k, f.generalizedEigenspace μ k ≠ ⊥} := by
```
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Eigenspace/Basic.lean:338]: totally optional:
```suggestion
    InjOn (⨆ k, f.generalizedEigenspace · k) {μ | ⨆ k, f.generalizedEigenspace μ k ≠ ⊥} := by
```

**Review by j-loreaux — COMMENTED** [review]
Looks pretty nice! Minor comments.

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean (hunk 2 above)
Comment (ocfnash) [reply Mathlib/LinearAlgebra/Eigenspace/Basic.lean]: Good point; in fact this already exists as `Algebra.mul_sub_algebraMap_commutes` so I have dropped `hc₁` and `hc₂` in favour of that.

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean (hunk 3 above)
Comment (ocfnash) [reply Mathlib/LinearAlgebra/Eigenspace/Basic.lean]: It saves a line break so I'm all in favour!

File: Mathlib/LinearAlgebra/Eigenspace/Basic.lean (hunk 1 above)
Comment (ocfnash) [reply Mathlib/LinearAlgebra/Eigenspace/Basic.lean]: Good point; in fact this already exists as `Algebra.mul_sub_algebraMap_commutes`.

**Comment by ocfnash** [comment]
bors merge

**Pull request closed without merging**